### PR TITLE
Remove `check_checkpoint_callback`

### LINKED ISSUE #6672: [RFC] Training Loop Checkpoint Consolidation
## 🚀 Feature
Currently, in training loop, at end of each training epoch and end of training, we checkpoint based on hook `on_validation_end`, which could be confusing. We would like to move those logics to dedicated hook. 

### Motivation

1. currently, at end of each training epoch (`run_training_epoch` in `training_loop.py`), we call model checkpoint callback and early stopping callback explicitly when there is no validation stage
```
if should_train_only:
     self.check_checkpoint_callback(True)
     self.check_early_stopping_callback(True)
```
however, we call the `on_validation_end` hook
```
  for cb in callbacks:
      cb.on_validation_end(self.trainer, model)
```
This is confusing. For example, for `ModelCheckpoint` callback, `on_validation_end` is dependent on parameter `every_n_val_epochs`. When there is no validation stage, triggering relied on `every_n_val_epochs` is not ideal. 

**Instead, we could implement a specific hook for controlling, which is independent of  `every_n_val_epochs`**

2. In addition, at end of training, we also call model checkpoint callback explicitly which again relies on `callback.on_validation_end`
Similarly, we should move this to hook `on_train_end`, and trigger should be independent of `every_n_val_epochs`.

### Pitch

**For 1.  inside `run_training_epoch`**

at the end of training epoch, instead of doing such
```
  if should_train_only:
      self.check_checkpoint_callback(True)
      self.check_early_stopping_callback(True)
```

do
```
if should_train_only:
   self.trainer.call_hook("on_train_epoch_without_validation_end")
```
Note: why need a new hook instead of reusing the existing one `on_train_epoch_end`,  as EarlyStoping and ModelCheckpoint are relying on `log_train_epoch_end_metrics ` and `update_learning_rates`. As this hook is called in the training epoch when no validation is enabled, propose to name as `on_train_epoch_without_validation_end`.

Implement `on_train_epoch_without_validation_end` for `ModelCheckpoint` and `EarlyStopping` callback.

**Trigger Change**:
* as it is triggered when validation is not enabled, instead of relying on `every_n_val_epochs`, we do it for every training epoch

**For 2. inside `on_train_end`**

at the end of whole training, remove
```
  # trigger checkpoint check. need to temporarily decrease the global step to avoid saving duplicates
  # when a checkpoint was saved at the last step
  self.trainer.global_step -= 1
  self.check_checkpoint_callback(should_update=True, is_last=True)
  self.trainer.global_step += 1
```
instead rely on hook
```
  # hook
  self.trainer.call_hook("on_train_end")
```
Implement `on_train_end` for `ModelCheckpoint`

**Trigger Change**:
* instead of relying on `every_n_val_epochs`, we provide option `trigger_on_train_end` to determine whether checkpoint.
By default, we will make it **off**. 

Note: why we default to **off**, as if we turn it on, there is a risk of checkpointing error for the following scenario:
* in checkpoint, we set monitor value to be some validation metric, like `val_loss` or `val_auc`, when training failure in the middle, and no validation has been ran before the failure (for example, user configure to check val for large number batches), it would complain that there is no valid monitor key and hence the model is not check-pointed. 
Once we turn it on, we could allow not validating the key for the end of the training. This way, when the monitor key is missing, though we will not be able to save top k, we could still save last (if save_last is turned on).

### Alternatives

For model checkpoint, instead of checkpoint at end of every training epoch, we introduce something like `every_n_train_epochs`, and check trigger condition based on `every_n_train_epochs`.

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Changes:
- Replace `training_loop.on_train_end` call to `check_checkpoint_callback` with the `on_train_end` hook implementation in `ModelCheckpoint`
- Always rely on the `ModelCheckpoint` hooks to save.

This resolves a bug where an extra checkpoint was saved at the end of training if the `val_check_interval` did not align with the training batches. In that case, a checkpoint was always saved as if `save_last` was set to `True` even if it was not. This change is reflected in the tests.

Fixes #6672

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: tests/checkpointing/test_model_checkpoint.py [hunk 1]
```diff
@@ -247,14 +247,7 @@ def configure_optimizers(self):
     ckpt_files = list(Path(tmpdir).glob('*.ckpt'))
     lr_scheduler_debug = trainer.dev_debugger.saved_lr_scheduler_updates
 
-    # on_train_end ckpt callback is called which creates an additional ckpt in case no ckpt is created at the
-    # end of epoch, thus if val_check_interval doesn't align with the training steps we create an additional ckpt
```
Comment (awaelchli) [inline tests/checkpointing/test_model_checkpoint.py:251]: does this not apply anymore because we have a check `_should_skip_saving_checkpoint` in the `ModelCheckpoint. on_train_end` ?

File: tests/checkpointing/test_checkpoint_callback_frequency.py [hunk 2]
```diff
@@ -50,7 +50,7 @@ def test_mc_called(tmpdir):
 @mock.patch('torch.save')
 @pytest.mark.parametrize(
     ['epochs', 'val_check_interval', 'expected'],
-    [(1, 1.0, 1), (2, 1.0, 2), (1, 0.25, 4), (2, 0.3, 7)],
+    [(1, 1.0, 1), (2, 1.0, 2), (1, 0.25, 4), (2, 0.3, 6)],
```
Comment (awaelchli) [inline tests/checkpointing/test_checkpoint_callback_frequency.py:53]: `save_last` is not set True, therefore we have on less checkpoint?

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -317,6 +317,23 @@ def on_validation_end(self, trainer: 'pl.Trainer', pl_module: 'pl.LightningModul
...
+        """
+        Save a checkpoint when training stops.
+
+        This will only save a checkpoint if `save_last` is also enabled as the monitor metrics logged during
+        training/validation steps or end of epochs are not guaranteed to be available at this stage.
+        """
+        if self._should_skip_saving_checkpoint(trainer):
+            return
+        if self.save_last and self.verbose:
+            rank_zero_info("Saving latest checkpoint...")
```
Comment (tchaton) [inline pytorch_lightning/callbacks/model_checkpoint.py:330]:
```suggestion
        if self._should_skip_saving_checkpoint(trainer) or not self.save_last:
            return
        if self.verbose:
            rank_zero_info("Saving latest checkpoint...")
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -317,6 +317,23 @@ def on_validation_end(self, trainer: 'pl.Trainer', pl_module: 'pl.LightningModul
             return
         self.save_checkpoint(trainer)
 
+    def on_train_end(self, trainer: 'pl.Trainer', pl_module: 'pl.LightningModule') -> None:
```
Comment (tchaton) [inline pytorch_lightning/callbacks/model_checkpoint.py:320]: Neat !

File: tests/checkpointing/test_checkpoint_callback_frequency.py (hunk 2 above)
Comment (tchaton) [reply tests/checkpointing/test_checkpoint_callback_frequency.py:53]: Can we parametrize `save_last=True` as we should have the same number of checkpoint as previously generated.

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:330]: This is not necessary as we have the check below but I guess it doesn't hurt

File: tests/checkpointing/test_model_checkpoint.py (hunk 1 above)
Comment (carmocca) [reply tests/checkpointing/test_model_checkpoint.py:251]: No, as `check_checkpoint_callback` called `on_validation_end` which also has `should_skip_saving_checkpoint`.

The reason is that the global step has increased when we check to save `on_train_end` when not aligned, and since the `on_validation_end` hook is used which does not early exit if `save_last=False`, it saves another checkpoint.

So we were forcing an extra "last" checkpoint which was not technically a `save_last` checkpoint

File: tests/checkpointing/test_checkpoint_callback_frequency.py (hunk 2 above)
Comment (carmocca) [reply tests/checkpointing/test_checkpoint_callback_frequency.py:53]: > save_last is not set True, therefore we have on less checkpoint?

Correct

> Can we parametrize save_last=True as we should have the same number of checkpoint as previously generated.

Will do, but this increases the expected count for all cases. The difference here is due to an implementation bug when the number of batches does not align with the validation check interval requested (see my other comment below)

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by awaelchli — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
